### PR TITLE
Allow loopback interfaces on all node types when specified by the user

### PR DESCRIPTION
**Pull request by ipspace**
This is the main logic for fixing #2261 and #2266. @jbemmel: would appreciate if you could double-check it.

Next, I have to create test cases to validate the edge conditions (**loopback.ipv4** on host, **loopback: True** and **loopback: False** work correctly), and finally run all known devices through these test cases to verify that the configuration templates don't break.

### CODE REVIEW

File: netsim/roles/router.py [hunk 1]
```diff
@@ -12,44 +12,5 @@
 
 from . import select_nodes_by_role
```
Comment (jbemmel) [inline netsim/roles/router.py:13]: No longer used :)

File: netsim/augment/nodes.py [hunk 2]
```diff
@@ -194,6 +194,61 @@ def check_duplicate_mgmt_addr(topology: Box) -> None:
...
+  lb_value = n.get('loopback',None)
+
+  if lb_value is False:                           # If 'loopback' is set to false, the user doesn't want it
+    n.pop('loopback',None)                        # ... so remove any loopback data
+    return                                        # ... and get out
+  elif lb_value is True:                          # Or maybe the user is adamant he wants a loopback?
+    n.loopback = {}                               # ... we can do that, will start with no extra data
+  elif lb_value is None:                          # Oh, the user doesn't care
+    if n.get('role','router') != 'router':        # ... so they will not get a loopback for a non-router device
+      return
```
Comment (jbemmel) [inline netsim/augment/nodes.py:210]: What about ```loopback: bla```?

File: netsim/augment/links.py [hunk 3]
```diff
@@ -1077,15 +1077,11 @@ def copy_link_gateway(link: Box, nodes: Box) -> None:
         intf.gateway[af] = link.gateway[af]
 
 """
-Set node.af flags to indicate that the node has IPv4 and/or IPv6 address family configured
+Set node.af flags to indicate that the node has IPv4 and/or IPv6 address family configured on its interfaces
 """
 def set_node_af(nodes: Box) -> None:
   for n in nodes.values():
     for af in ['ipv4','ipv6']:
```
Comment (jbemmel) [inline netsim/augment/links.py:1084]: ```for af in ['ipv4','ipv6'] if af not in n.af```

**Review by jbemmel — CHANGES_REQUESTED** [review]
Couple of corner cases

File: netsim/augment/nodes.py [hunk 4]
```diff
@@ -194,6 +194,61 @@ def check_duplicate_mgmt_addr(topology: Box) -> None:
...
+  pool = n.get('loopback.pool','loopback')
+  prefix_list = addressing.get(pools,[ pool ],n.id)
+
+  for af in prefix_list:
+    if prefix_list[af] is True:
+      log.error(
+        f"Address pool {pool} cannot contain unnumbered/LLA addresses",
+        category=log.IncorrectType,
+        module='nodes')
+    elif not n.loopback[af] and not (prefix_list[af] is False):
```
[the remark this answers is not in the document]
Comment (ipspace) [reply netsim/augment/nodes.py:229]: Thank you, will check and add to a test case

File: netsim/augment/links.py (hunk 3 above)
Comment (ipspace) [reply netsim/augment/links.py:1084]: Yeah, we could do that optimization. Won't buy us much, but it's nice to have ;)

**Comment by ipspace** [comment]
@jbemmel: Addressed all your corner cases (thank you), made loopback parameter processing as flexible as I possibly could, documented the changes, fixed the templates that broke without the loopback interface...

You might want to look at this again before I merge it.

File: tests/topology/expected/dual-stack.yml [hunk 5]
```diff
@@ -238,7 +238,6 @@ nodes:
     loopback:
       ifindex: 0
       ifname: Loopback0
-      ipv4: 10.0.0.1/32
```
Comment (jbemmel) [inline tests/topology/expected/dual-stack.yml:241]: Is this expected? Topology has dual-stack loopback pool

File: tests/topology/expected/unnumbered.yml [hunk 6]
```diff
@@ -36,12 +36,10 @@ links:
   - ifindex: 2
     ifname: Ethernet1/2
     ipv4: true
-    ipv6: true
```
Comment (jbemmel) [inline tests/topology/expected/unnumbered.yml:39]: Expected?

**Review by jbemmel — APPROVED** [review]
looks ok to me
